### ISSUE TITLE
Large-scale performance and formatting tests

### ISSUE THREAD

**Issue by AaronDavidNewman** [vexflow/vexflow#38]
I've been enhancing the reverse-engineering capabilities of Smoosic, so that it can output vex code that runs independently of any application (including Smoosic), and with different versions of vexflow.  I think it fills a void in our testing capabilities , both in terms of scale, and also in the sense that our unit tests aren't very realistic, musically (100s of clefs, etc.)

I don't think these should go in the vexflow repository.  It will be too large - each example is 1000s  of lines and there will be 100-ish different tests.  I think it makes sense to have a different repository that can interact with different versions of vexflow, and also create a useful (if not realistic) performance comparison.

Anyway, here are a couple of examples. 
https://jsfiddle.net/AaronDavidNewman/eoucfL16/

https://jsfiddle.net/AaronDavidNewman/ykx0h76a/7/

**Comment by sschmidTU** [comment]
Great idea!
Unrelated comment: I would y-align lyrics (set each to the lowest y-position in the staffline), otherwise they're unpleasant to read.

**Comment by rvilarl** [comment]
I think also that it is a good idea. What is the source of the text musicxml?

**Comment by AaronDavidNewman** [comment]
> I think also that it is a good idea. What is the source of the text musicxml?

Most of these are either my arrangements, transcriptions, or things I've manually entered in Smoosic to test it out (e.g. GFH Messiah).

I can import music xml (and midi) if you have specific things you think should be represented.  Smoosic has its own serializable format (that's the SMO part of the name).

**Comment by rvilarl** [comment]
@AaronDavidNewman there was a discussion about making tests with larger scores https://github.com/0xfe/vexflow/issues/441 perhaps you could provide some of those.

**Comment by AaronDavidNewman** [comment]
Here is what I have so far.  There is some Joplin, some Bach, some stuff with lyrics, voices, ornaments etc.  Going to add some  Handel and some more instrumental stuff.   At the current time there are 15-20 different vex files, and  6 versions of the vexflow library.

Web interface is very "web 1.0" right now.  Need to add some exception handling and error reporting and I'll add some QUnit-style css.  

When it is better we can fork it into a vexflow project page.

There are some rendering issues:

* 4.0.3 doesn't render anything that uses Leland font.
* Almost nothing renders with the early version of vexflow 5.  Errors in annotation, chord changes and ornaments.
* Some more minor issues with beaming/slur directions that happen different than in Smoosic, not quite sure why yet.

Project web page:
https://aarondavidnewman.github.io/VexRender/index.html

Project source:
https://github.com/AaronDavidNewman/VexRender

**Comment by rvilarl** [comment]
> Almost nothing renders with the early version of vexflow 5. Errors in annotation, chord changes and ornaments.

have you loaded the fonts?

**Comment by AaronDavidNewman** [comment]
I have updated the menus now, so you can switch libraries or scores to see how the same score looks in different libraries.

https://aarondavidnewman.github.io/VexRender/html/BachWTC-0422.html

![image](https://github.com/vexflow/vexflow/assets/5438280/0bf8a2d1-f7f8-47d6-8662-5898f64cbbb0)

**Comment by rvilarl** [comment]
@AaronDavidNewman Thanks, this helps a lot! I will look into why vexflow5 is being so much slower. It is surprising taking into account the ammount of porting done. Perhaps it is the time taken to download the fonts. I will invetigate.

**Comment by ronyeh** [comment]
I just saw this.

WOW.

Thanks @AaronDavidNewman for this huge undertaking.

I think it will be a wonderful resource and we should definitely link to your test site from our repo / docs at some point.

And having you work on this will be a great way for Rodrigo and I to make sure we don't break every single API from V4, hahah. At least there will be one developer (you) trying to upgrade from V4 to V5, and telling us what things we have done are causing you extra pain. :-)

**Comment by ronyeh** [comment]
> @AaronDavidNewman Thanks, this helps a lot! I will look into why vexflow5 is being so much slower. It is surprising taking into account the ammount of porting done. Perhaps it is the time taken to download the fonts. I will invetigate.

YES, a huge block during the initial load is the downloading of all the web fonts.

I have a work in progress PR #61 that will help with local testing, since all the fonts will be pre-installed into node_modules.

However, for a real deployment, only ONE or TWO fonts should be preloaded... otherwise it will make the app too slow.

**Comment by AaronDavidNewman** [comment]
> YES, a huge block during the initial load is the downloading of all the web fonts.

Maybe, but I don't think that the only issue.  I ran this with cache enabled, so fonts should have been cached.  And also, if the issue were a 1-time load overhead, I would expect the same delta to be the same for each file.  It appears that the increase is proportional to the overall rendering time.

In looking at the performance breakdown by function, the increase in time between vex5 and vex4 is modest but pervasive.   The page load has completed before the main function starts.  The time spent in the draw functions/svg context functions is greater.  This means either they are taking longer, or they are being called more frequently.

I think I will need to create a vex5 version of smoosic to get a good formatting comparison.  Since the variable names have changed, the beam direction and other formatting parameters are wrong for vex5.  I don't know if that affects the render time - it seems unlikely but who knows.  

Call tree for vex 5:
![image](https://github.com/vexflow/vexflow/assets/5438280/d2545e83-1e01-4581-a9a6-9726631c6460)

Call tree for vex 4:
![image](https://github.com/vexflow/vexflow/assets/5438280/bb41dda2-64d2-4885-9977-839e6d3a469a)

**Comment by rvilarl** [comment]
I believe that we have to wait until more classes is ported to look into the details. As a minimum the notes and accidentals.

**Comment by ronyeh** [comment]
Rodrigo, do you think the calls to measureText() could be a contributing
factor to the worse performance?

That's the one thing I noticed as I have been reviewing the PRs. Every time
we change the text on a glyph we also have to call measureText(). This
would scale linearly with the number of glyphs rendered.

**Comment by rvilarl** [comment]
Hi Ron, the complete vexflow5 prototype was quicker than VexFlow4. No doubt that we have to minimise the calls to measureText and I am sure that this has a big weight in performance.

**Comment by AaronDavidNewman** [comment]
I have updated the render tests to use different code for v4 vs. v5.  This fixes a lot of the issues, e.g. the beam directions since I am not outputting the correct identifiers.  Chord symbols still look wrong and the render is a bit slower, but most renders look almost exactly the same between v4 and v5.

https://aarondavidnewman.github.io/VexRender/html/Solovey-0vexv5.html

**Comment by rvilarl** [comment]
Thanks @AaronDavidNewman, what version 5 is used? The early package released by @ronyeh ?

**Comment by ronyeh** [comment]
I can make another early access release tomorrow evening. 

Rodrigo, do you have a version of your app running on your **fully migrated** vexflow 5? It would be awesome to be able to view yours and Aaron's sites to see what v5 is good at, and what can be improved.

**Comment by rvilarl** [comment]
The complete draft is no longer in line with the changes. I recommend the accidental branch.

**Comment by rvilarl** [comment]
The complete draft is no longer in sync with the changes. I recommend the accidental branch.

**Comment by AaronDavidNewman** [comment]
I used the master as of yesterday, and created my own package.

**Comment by ronyeh** [comment]
For anyone else who wants to quickly test the current alpha, you can do:

```
npm install vxflw-early-access@5.0.0-alpha.1
```

I just released the most recent main revision:
https://github.com/vexflow/vexflow/issues/13#issuecomment-1695033448

**Comment by rvilarl** [comment]
> It's explained pretty clearly here:
> https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/pull/1139
> Vexflow 3+ has layout issues, since they introduced a brand new formatter (by 0xfe), which was not the focus of improvements (though great improvements came in other parts), and shortly after 0xfe became inactive. The current best version for OSMD is 1.2.93. It produces the best results overall. It would be a lot of work to fix all the layout issues with newer Vexflow versions, and additionally adapting to all the changes, and porting our custom Vexflow patches. >Also, Vexflow 5 is currently breaking a lot of older APIs, and they're not done breaking them yet, so I'd prefer to wait until most of that is finished.

_Originally posted by @sschmidTU in https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/discussions/1453#discussioncomment-6865659_

@AaronDavidNewman would it be easy to add the library version 1.2.93?

**Comment by AaronDavidNewman** [comment]
> @AaronDavidNewman would it be easy to add the library version 1.2.93?

I'm afraid not.  Even 4.0.3 doesn't work for all cases.  For reasons, one is Smoosic uses Vexflow types and 1.2.93 didn't even have types.

It might work to create the OSMD problem scores in Smoosic and have Smoosic generate the Vexflow to produce a generic case.   Since I think OSMD gets its input from Music XML, we could import those same XML files into Smoosic.  They will have to be tweaked some (e.g. slur directions don't work the same in Smoosic as MusicXML), but that work would be minimal and I can help with that. 

I really feel like the OSMD formatting problems are minor and can be addressed, but since the issues exist in the context of OSMD it's been hard to extract the problems to Vexflow test cases,

**Comment by rvilarl** [comment]
@AaronDavidNewman could you update to the latest v5. I would like to start with the profiling.

**Comment by AaronDavidNewman** [comment]
This is done.

A couple of new artifacts.  In TextBracket:

![image](https://github.com/vexflow/vexflow/assets/5438280/01d9a5f3-d158-4796-bf24-19041242ac31)

from `const smo253077 = new VF.TextBracket({ start: smo242249, stop: smo242279, text: 'ritard', position: 1 });`

similar issue in ChordSymbol:
![image](https://github.com/vexflow/vexflow/assets/5438280/4695a8ff-05f0-4d28-b58d-5be20019f0ab)

**Comment by rvilarl** [comment]
@AaronDavidNewman thanks a lot

**Comment by rvilarl** [comment]
> This is done.
> 
> A couple of new artifacts. In TextBracket:
> 
> ![image](https://user-images.githubusercontent.com/5438280/268524915-01d9a5f3-d158-4796-bf24-19041242ac31.png)
> 
> from `const smo253077 = new VF.TextBracket({ start: smo242249, stop: smo242279, text: 'ritard', position: 1 });`
> 
> similar issue in ChordSymbol: ![image](https://user-images.githubusercontent.com/5438280/268524988-4695a8ff-05f0-4d28-b58d-5be20019f0ab.png)

@AaronDavidNewman the artifact is related to a wrong configuration of the stack of fonts. 
```js
VF.setMusicFont("Bravura","Gonville","Custom");
```

Could you change that to:
```js
VF.setMusicFont("Bravura","Roboto Slab");
```

**Comment by rvilarl** [comment]
The speed of the VexFlow 5 is better to the previous one!!! :)

**Comment by AaronDavidNewman** [comment]
I can do this, and it works but there is some overlap.  It seems to happen regardless of the size.  Using setMusicFont is global, but you might want different text fonts for e.g. tempo vs. textbracket vs annotation, etc.

``` typescript
smo253256.setFont({ family: 'Roboto Slab', size: 12, style: 'italic' });
```

![image](https://github.com/vexflow/vexflow/assets/5438280/5e6b81e4-6203-45c6-a2ac-53571a96bd6e)

**Comment by AaronDavidNewman** [comment]
> The speed of the VexFlow 5 is better to the previous one!!! :)

Oddly enough, the Bach WTC is now the slowest, relative to Vex 4.   I should create a histogram thing so we can get an idea of changes over time.

**Comment by rvilarl** [comment]
> I can do this, and it works but there is some overlap. It seems to happen regardless of the size. Using setMusicFont is global, but you might want different text fonts for e.g. tempo vs. textbracket vs annotation, etc.
> 
> ```ts
> smo253256.setFont({ family: 'Roboto Slab', size: 12, style: 'italic' });
> ```
> 
> ![image](https://user-images.githubusercontent.com/5438280/268532428-5e6b81e4-6203-45c6-a2ac-53571a96bd6e.png)

You can do that in two different ways one is as you did for a specific element and a second one  is in CommonMetrics for the class. We are going to export this class as Metrics. see #78

**Comment by AaronDavidNewman** [comment]
I updated the rendering library with v5 alpha 11.  The results and formatting is about the same.  Well-tempered clavier is still slower in v5.

https://aarondavidnewman.github.io/VexRender/html/BachWTC-2vexv5.html

**Comment by rvilarl** [comment]
@AaronDavidNewman thanks for that. There is something completely wrong with the beaming as well. Look at measure 6. Vexflow5 is presenting it wrong compared to 422

**Comment by AaronDavidNewman** [comment]
The measures for the pages won't exactly line up, since it seems Vex5 is actually a bit more compact and so more measures fit on a page.  Also, Smoosic uses its own beaming logic, so if there is something wrong with the beaming it is probably a Smoosic issue.

In fact, that may explain the rendering difference, since v5 will is rendering more measures for this piece.  I'm not sure why, it's also possible I need to regenerate the V4 versions, there have been some formatting changes in Smoosic.

**Comment by rvilarl** [comment]
You are right vexflow5 is showing 34 and vexflow4 26. That might be good the difference.

**Comment by rvilarl** [comment]
@AaronDavidNewman now that we are about to release I wanted to ask you if you would like to place vexRender in vexflow organization. I think it is a fantastic tool to keep and eye on performance. If you do not mind we can also simply fork it here and maintain the different versions of Vexflow.
